### PR TITLE
Duplication of ASTs, in support of upcoming script optimization changes

### PR DESCRIPTION
**Pull request by vpax**
Here's a standalone version of the branch first discussed in https://github.com/zeek/zeek/pull/1255.  It doesn't add any user-visible functionality.  The functionality can be tested using a new `-a AST-dup` btest alternative (requires Robin's `btest` branch that supports multiple baselines).

### CODE REVIEW

File: src/Stmt.cc [hunk 1]
```diff
@@ -110,6 +111,11 @@ bool Stmt::IsPure() const
 	}
 
 void Stmt::Describe(ODesc* d) const
+	{
+	StmtDescribe(d);
```
Comment (rsmmr) [inline src/Stmt.cc:115]: Why the rename of all the derived methods when the new top-level just does a direct call to the same thing?

File: src/Expr.h [hunk 2]
```diff
@@ -226,6 +231,9 @@ class NameExpr final : public Expr {
 
 	TraversalCode Traverse(TraversalCallback* cb) const override;
 
+	// Optimization-related:
```
Comment (rsmmr) [inline src/Expr.h:234]: I know this has been discussed in the previous PR, but I'm still confused: is this functionality optimizer-specific or not? I thought the is answer is "no", but the consistent comments still leave it ambiguous.

File: src/Stmt.h [hunk 3]
```diff
@@ -66,10 +67,13 @@ class Stmt : public Obj {
 
 	virtual TraversalCode Traverse(TraversalCallback* cb) const = 0;
 
+#include "script_opt/StmtOpt-Public.h"
```
Comment (rsmmr) [inline src/Stmt.h:70]: Not a fan of this (and similar pieces elsewhere). Just flagging for now, need to think about other options.

**Comment by rsmmr** [comment]
> can be tested using a new `-a AST-dup` btest alternative (requires Robin's `btest` branch that supports multiple baselines).

(Ha, didn't realize you were going to use it for this one already (vs. the full script-opt later)). 😄 )

This seems quite heavy-handed for such an internal change, requiring a full 2nd test/CI run. I agree that we should test this code, especially given that we'd be merging without it being used anywhere. I'm wondering if we can test this more directly through C++-side unit tests. I believe doing this on a per-class basis would be sufficient; I don't think we need to necessarily test it with large complex ASTs (if each class can correctly duplicate itself, we can trust recursion to do the rest). So how about we write one(*) unit test per class that (1) instantiates a small AST node of that kind with just basic child nodes, (2) duplicates that, and (3) uses traversal on both instances to see if the result are the same. 

((*) or a small number, if an AST has variants leading to different structures)

File: src/Stmt.cc (hunk 1 above)
Comment (vpax) [reply src/Stmt.cc:115]: The statement body used to read:
``` 
        if ( d->DoOrig() )
                Original()->StmtDescribe(d);
        else
                StmtDescribe(d);
```
but @jsiwek was not a fan of `ODesc` knowing about the notion of "original".  I'd like to leave this factoring in place, though, because I'm expecting that something like this (= wrapping `Stmt` descriptions with some refinements) will be needed once we progress to the AST transformations that are in the future.

File: src/Expr.h (hunk 2 above)
Comment (vpax) [reply src/Expr.h:234]: Well, yes & no.  Yes in that it's only used by script optimization.  No in that it doesn't "know" anything about how script optimization works.  The next round of changes are going to add more methods flagged `Optimization-related`, so we'll have that section in the class definitions anyway (modulo resolution of your point below).

File: src/Stmt.h (hunk 3 above)
Comment (vpax) [reply src/Stmt.h:70]: FYI, I tried going the multiple inheritance route that you'd suggested.  There's a comment at the beginning of `script_opt/ExprOpt-Public.h`: https://github.com/zeek/zeek/blob/96848687c98eaaa4908834dee26737ec6f3d5e53/src/script_opt/ExprOpt-Public.h#L5 of why I concluded it would be quite a bit uglier still.  The core problem is that the `Stmt` and `Expr` subclasses need to override the optimization-related virtual methods, which means the change is actually more invasive.

I also gave thought to whether this could be done using accessors/mutators, but ultimately that's going to require a lot of massive switch statements, which seems to me antithetical to object-oriented programming.

**Comment by vpax** [comment]
> I agree that we should test this code, especially given that we'd be merging without it being used anywhere. I'm wondering if we can test this more directly through C++-side unit tests.

I think of this differently.  This change, and a bunch of others in the pipeline, are all providing _intermediary_ functionality.  They're not meant to be actually used in isolation.  Their functionality will be tested collectively, and end-to-end, by using CI for compiled code (all stages active).  Only when that CI fails will the maintainer want to determine which stage is having problems; for that they can run the `btest`'s using `-a` alternatives to isolate the problem.

File: src/Stmt.h (hunk 3 above)
Comment (jsiwek) [reply src/Stmt.h:70]: I may have lost track of some larger goals / problem-definitions, so sorry for potentially adding noise/confusion but I'm not sure it's worthwhile to jump through the script-optimization demarcation hoops for this case:

* Even though it's practically expected to only be used by script-opt logic, AST-duplication is conceptually independent and that's what allows it to be a candidate for independent merging in the first place, so I'm confused about the attempt to still draw boundaries that put it within script-opt territory.  Can't we simply add methods to Expr.h/Stmt.h in this instance ?

* Both `Expr` and `Stmt` are in the `detail` namespace now and that's probably a sufficient way to communicate an interface (in)stability disclaimer to users.  Also not to conflate too much with release timelines, but certainly after Zeek 4.0, there's an awareness of this and we might be free to do more significant changes/re-org to `Expr` / `Stmt` if it helps solve problems in whatever way is most efficient rather than potentially being distracted or misled by a pre-orientation towards objects.

File: src/Var.cc [hunk 4]
```diff
@@ -701,8 +701,20 @@ TraversalCode OuterIDBindingFinder::PostExpr(const Expr* expr)
...
 	{
+	if ( duplicate_ASTs && reporter->Errors() == 0 )
+		// Only try duplication in the absence of errors.  If errors
+		// have occurred, they can be re-generated during the
+		// duplication process, leading to regression failures due
+		// to duplicated error messages.
+		//
+		// We duplicate twice to make sure that the AST produced
+		// by duplicating can itself be correctly duplicated.
+		body = body->Duplicate()->Duplicate();
```
Comment (jsiwek) [inline src/Var.cc:716]: A double-duplication seems more like a single test-case than something that has to always be done in reality.  Maybe that's exactly the thinking here since AST-duplication would essentially be unused at this point except as a hypothetical `btest -a AST-dup` run, but seems like an inefficiency that's easy to accidentally leave in later on.

**Comment by jsiwek** [comment]
I must have forgot already: what was the benefit of merging this particular piece independently?

Don't think I have much conceptual concern anymore for it getting merged early (even in terms of testing/CI stuff since it's basically all in `detail` and will definitely end up receiving any necessary test-coverage/fixes later on anyway), but also could see this just being naturally organized as a single commit within a larger PR, one that will actually end up making good use of the new duplication functions.

File: src/Stmt.h (hunk 3 above)
Comment (vpax) [reply src/Stmt.h:70]: > I'm not sure it's worthwhile to jump through the script-optimization demarcation hoops for this case

I agree, but right behind it I have the changes for inlining, and that _will_ add methods that are specific to script optimization.

I really don't care about this one either way, was just trying to lay down a framework that would work going forward, given @rsmmr earlier commented that the optimization stuff should be well-delineated.

> * Both `Expr` and `Stmt` are in the `detail` namespace now and that's probably a sufficient way to communicate an interface (in)stability disclaimer to users

Note that there's something like 4,000 LOC coming as additions to `Expr.cc`/`Stmt.cc`, hence the desireability of somehow separating it out ...

File: src/Var.cc (hunk 4 above)
Comment (vpax) [reply src/Var.cc:716]: > A double-duplication seems more like a single test-case than something that has to always be done in reality

Yep.  But if it succeeds then you know that a single duplication will succeed.

> Maybe that's exactly the thinking here since AST-duplication would essentially be unused at this point except as a hypothetical `btest -a AST-dup` run

Right.

> seems like an inefficiency that's easy to accidentally leave in later on

You can only get to that code path by setting a flag that's off by default, and the use of double-duplicating is documenting, so IMHO it's safe enough in that regard.  That said, all I really want at this point is a prospect of moving forward with the whole-lotta-changes to come ...

**Comment by vpax** [comment]
> I must have forgot already: what was the benefit of merging this particular piece independently?

It was just to keep things bite-sized for review (or at least closer to bite-sized).  I'm happy to defer it to adding inlining if that's desired.  Or, put differently: if this is integrated, I can have the inlining draft PR ready quite soon, so there will only a brief interval when this is in but that one isn't.

**Comment by rsmmr** [comment]
Backing up a bit here, looks like we need to approach this differently. I suggested individual merges originally in the hope that we could get to bite-sized pieces that are easy to review/refine standalone. I  generally still think that that's a viable strategy even for pieces that aren't immediately used. However, a change isn't "standalone" in this sense if the motivation for doing things a certain way is purely residing with future work that's going to come separately. The problem with that is that one cannot really judge it then by looking just at the PR under review (which is of course the whole point of being "bite-sized"). The "Val union" change is an instance where the "standalone" assumption seems to be working, but for this one it isn't if we can't consider "AST duplication" as a general mechanism outside of the script-opt branch (I still think  *conceptually* we could, but looks like the implementation comes with sufficient script-opt specifics that reviewing gets stuck there). 

The other way to approach this is making the PRs larger by including enough functionality so that it does become "standalone" in the above sense. However, we need to be careful that the review remains manageable. What tends to work best is what Jon mentioned above: creating a branch that stacks individual commits so that each one puts a specific piece in place that can still be understood/reviewed on its own, but with the difference being that now it's ok to point to subsequent commits inside that same PR as explanation for why something looks a certain way (and to make that explicit in the commit message). The smaller each commit in that stack, the better. A challenge with this approach is that it's harder to iterate because it will require changing individual commits inside the stack as feedback gets integrated (usually leading to `git rebase -i && git push --force` cycles).

**Comment by rsmmr** [comment]
> Here, I'm wondering whether it would suffice to stack the AST duplication with inlining

Summarizing a quick offline discussion: we'll try this. The goal is to come to a PR that doesn't need much additional context about what's coming later. We may or may to stack it further, we'll see.

**Comment by vpax** [comment]
I've now fixed the merge issues and opened a separate draft PR, #1303.
